src/Cache/Server/Server.php

```diff
@@ -0,0 +1,576 @@
+
+		foreach ( $this->spawn_queue as list( $key, $time_to_live, $is_network ) ) {
+
+			if ( $this->is_spawning( $key, $is_network ) ) {
+				continue;
+			}
+
+			$is_network ? $network_keys[] = $key : $keys[] = $key;
+
+			$requests[ $key ] = [
+				'url'     => home_url(),
+				'headers' => [
+					self::HEADER_KEY => $key,
+					self::HEADER_TTL => $time_to_live,
+				],
+				'data'    => '',
+				'cookies' => [],
+			];
+		}
+
+		$this->mark_spawning( true, ...$network_keys );
+		$this->mark_spawning( false, ...$keys );
+
+		\Requests::request_multiple(
+			$requests,
+			[
+				'timeout'          => 0.01,
+				'follow_redirects' => false,
+				'blocking'         => false,
+				'type'             => \Requests::HEAD,
+			]
+		);
+
+		return $requests;
+	}
+
+	/**
+	 * Use transients to mark the given key as currently being sent via an update request, to prevent
+	 * multiple concurrent request.
+	 * Transients will not be deleted manually, but are set with a very short expiration so they will expire and vanish
+	 * in few seconds when (hopefully) all the parallel-executing updating requests finished.
+	 *
+	 * @param bool     $is_network
+	 * @param string[] $keys
+	 */
+	private function mark_spawning( bool $is_network, string ...$keys ) {
+
+		foreach ( $keys as $key ) {
+			$is_network
+				? set_site_transient( self::SPAWNING_KEYS_TRANSIENT . md5( $key ), 1, 10 )
+				: set_transient( self::SPAWNING_KEYS_TRANSIENT . md5( $key ), 1, 10 );
+		}
+	}
+
+	/**
+	 * @param string $key
+	 * @param bool   $is_network
+	 *
+	 * @return bool
+	 */
+	private function is_spawning( $key, bool $is_network ): bool {
+
+		return $is_network
+			? (bool) get_site_transient( self::SPAWNING_KEYS_TRANSIENT . md5( $key ) )
+			: (bool) get_transient( self::SPAWNING_KEYS_TRANSIENT . md5( $key ) );
+	}
+
+	/**
+	 * @param callable $updater
+	 * @param array    $args
+	 * @param mixed    $current_value
+	 *
+	 * @return array
+	 */
+	private function fetch_updated_value( callable $updater, array $args, $current_value = null ) {
+
+		try {
+
+			array_unshift( $args, $current_value );
+			$value   = $updater( ...$args );
+			$success = true;
+
+		} catch ( \Throwable $throwable ) {
+
+			$value   = $current_value;
+			$success = false;
+		}
+
+		return [ $value, $success ];
+	}
+
+	/**
+	 * @param string $namespace
+	 * @param string $key
+	 */
+	private function bail_if_not_registered( string $namespace, string $key ) {
+
+		if ( $this->is_registered( $namespace, $key ) ) {
+			return;
+		}
+
+		/*
+		 * TODO: Custom exception class?
+		 */
+		throw new \OutOfRangeException(
+			sprintf(
+				'The namespace/key pair "%s"/"%s" does not belong to any registered cache logic in %s.',
+				$namespace,
+				$key,
+				__CLASS__
+			)
+		);
+
+	}
+
+	/**
+	 * Cancelable (via `remove_action`) plus test-friendly way to end the request.
+	 */
+	private function exit_on_update_end() {
+
+		add_action( 'multilingualpress.after_server_update_value', function () {
```

Comment:
Should we make the action name available through a class constant?

Reply:
I did not found any `mlp_exit()` function in the branch, so did not changed this.

Reply:
It's a namespaced function, `Inpsyde\MultilingualPress\exit`. I'll adapt the code.